### PR TITLE
Add parameterized read support to the engine

### PR DESCRIPTION
**Pull request by Frassle**
This enables the engine to receive ReadResource for parameterized providers.

### CODE REVIEW

File: pkg/engine/lifecycletest/testdata/output/TestReplacementParameterizedProvider/refresh/progress.stdout.txt [hunk 1]
```diff
@@ -5,13 +5,15 @@
  <{%reset%}>  <{%reset%}> pulumi:providers:pkgA default_1_0_0 <{%reset%}><{%reset%}> 
  <{%bold%}><{%fg 3%}>~ <{%reset%}> pkgA:m:typA resA <{%bold%}><{%fg 3%}>refreshing<{%reset%}> 
  <{%reset%}>  <{%reset%}> pkgA:m:typA resA <{%reset%}><{%reset%}> 
- <{%bold%}><{%fg 3%}>~ <{%reset%}> pulumi:providers:pkgExt default_0_5_0 <{%bold%}><{%fg 3%}>refreshing<{%reset%}> 
- <{%reset%}>  <{%reset%}> pulumi:providers:pkgExt default_0_5_0 <{%reset%}><{%reset%}> 
+ <{%bold%}><{%fg 3%}>~ <{%reset%}> pulumi:providers:pkgExt default_1_0_0 <{%bold%}><{%fg 3%}>refreshing<{%reset%}> 
+ <{%reset%}>  <{%reset%}> pulumi:providers:pkgExt default_1_0_0 <{%reset%}><{%reset%}> 
  <{%bold%}><{%fg 3%}>~ <{%reset%}> pkgExt:m:typA resB <{%bold%}><{%fg 3%}>refreshing<{%reset%}>
```
Comment (Frassle) [inline pkg/engine/lifecycletest/testdata/output/TestReplacementParameterizedProvider/refresh/progress.stdout.txt:10]: Huh this looks wrong, why'd the version change on refresh.

Comment (Frassle) [reply pkg/engine/lifecycletest/testdata/output/TestReplacementParameterizedProvider/refresh/progress.stdout.txt:10]: OK this is on master as well. Will fix in another PR then come back to this.

Comment (Frassle) [reply pkg/engine/lifecycletest/testdata/output/TestReplacementParameterizedProvider/refresh/progress.stdout.txt:10]: https://github.com/pulumi/pulumi/pull/16785

File: pkg/resource/deploy/source_eval_test.go [hunk 2]
```diff
@@ -660,11 +660,11 @@ func TestDisableDefaultProviders(t *testing.T) {
...
 				}
 				// Perform some reads and invokes with explicit provider references.
-				_, _, perr := resmon.ReadResource("pkgA:m:typA", "resA", "id1", "", nil, aPkgProvider, "", "")
+				_, _, perr := resmon.ReadResource("pkgA:m:typA", "resA", "id1", "", nil, aPkgProvider, "", "", "")
 				aErrorAssert(t, perr)
-				_, _, perr = resmon.ReadResource("pkgB:m:typB", "resB", "id1", "", nil, providerBRef.String(), "", "")
+				_, _, perr = resmon.ReadResource("pkgB:m:typB", "resB", "id1", "", nil, providerBRef.String(), "", "", "")
 				assert.NoError(t, perr)
-				_, _, perr = resmon.ReadResource("pkgC:m:typC", "resC", "id1", "", nil, "", "", "")
+				_, _, perr = resmon.ReadResource("pkgC:m:typC", "resC", "id1", "", nil, "", "", "", "")
```
Comment (lunaris) [inline pkg/resource/deploy/source_eval_test.go:667]: Given that we've not extended these tests to deal with parameterisation, is it fair to say that we are moving away from using them in favour of lifecycle tests? If so, is there a route to removing them down the road?

**Review by lunaris — APPROVED** [review]

File: pkg/resource/deploy/source_eval_test.go (hunk 2 above)
Comment (Frassle) [reply pkg/resource/deploy/source_eval_test.go:667]: I think these tests were added to try and increase coverage numbers without much thought as to what "feature" exactly was being tested. So yes, they ought to be rewritten as targeted lifecycle tests but I'm not sure what coverage this was giving us. 
Most of it is probably just error paths, which honestly probably still worth having in lifecycle.

**Pull request merged**